src/derivers/cip3.ts

```diff
@@ -0,0 +1,384 @@
+  const extension = isHardened
+    ? getKeyExtension(
+        Z_TAGS.hardened,
+        parentNode.privateKeyBytes,
+        childIndex + BIP_32_HARDENED_OFFSET,
+      )
+    : getKeyExtension(Z_TAGS.normal, parentNode.publicKeyBytes, childIndex);
+
+  // entropy = Fcp(extension)
+  const entropy = generateEntropy({
+    chainCode: parentNode.chainCodeBytes,
+    extension,
+  });
+
+  const zl = entropy.subarray(0, 32);
+  const zr = entropy.subarray(32);
+
+  const parentKl = parentNode.privateKeyBytes.subarray(0, 32);
+  const parentKr = parentNode.privateKeyBytes.subarray(32);
+
+  // 8[ZL] + kPL
+  const childKl = add(trunc28Mul8(zl), parentKl);
+  // ZR + kPR
+  const childKr = add(zr, mod2Pow256(parentKr));
+  return concatBytes([childKl, childKr]);
+};
+
+type DeriveWithoutPrivateArgs = DeriveKeyBaseArgs & {
+  parentNode: {
+    chainCodeBytes: Uint8Array;
+    publicKeyBytes: Uint8Array;
+  };
+  isHardened: false;
+};
+
+const CHAIN_CODE_TAGS = {
+  normal: 3,
+  hardened: 1,
+};
+
+/**
+ * Derives a child chainCode.
+ *
+ * Following "Section V. BIP32-ED25519: SPECIFICATION, C.3" in https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf.
+ *
+ * @param param1 - The parameters for deriving a child chainCode.
+ * @param param1.parentNode - The parent node containing optionally a private key, chain code, and public key.
+ * @param param1.childIndex - The index of the child key.
+ * @param param1.isHardened - Indicates if the child key is hardened.
+ * @returns The derived child chainCode.
+ */
+export const deriveChainCode = async ({
+  parentNode,
+  childIndex,
+  isHardened,
+}: DeriveWithPrivateArgs | DeriveWithoutPrivateArgs) => {
+  // extension = i >= 2^31 ? (0x01||kp||i) : (0x03||Ap||i)
+  const extension = isHardened
+    ? getKeyExtension(
+        CHAIN_CODE_TAGS.hardened,
+        parentNode.privateKeyBytes,
+        childIndex + BIP_32_HARDENED_OFFSET,
+      )
+    : getKeyExtension(
+        CHAIN_CODE_TAGS.normal,
+        parentNode.publicKeyBytes,
+        childIndex,
+      );
+
+  // entropy = Fcp(extension)
+  const entropy = generateEntropy({
+    chainCode: parentNode.chainCodeBytes,
+    extension,
+  });
+
+  return entropy.subarray(32);
+};
+
+const PUBLIC_KEY_TAGS = {
+  normal: 2,
+};
+
+type DerivePublicKeyArgs = DeriveWithoutPrivateArgs & {
+  curve: Cip3SupportedCurve;
+};
+
+/**
+ * Derives a public key.
+ *
+ * Following "Section V. BIP32-ED25519: SPECIFICATION, D" in https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf.
+ *
+ * @param param1 - The parameters for deriving a child public key.
+ * @param param1.parentNode - The parent node containing chain code, and public key.
+ * @param param1.childIndex - The index of the child key.
+ * @param param1.curve - Derivation curve.
+ * @returns The derived child public key.
+ */
+export const derivePublicKey = async ({
+  parentNode,
+  childIndex,
+  curve,
+}: DerivePublicKeyArgs) => {
+  // extension = (0x02||Ap||i)
```

Comment:
Just for clarification: The spec describes a way to derive public keys for `i >= 2^31`, but we don't need to implement that because we can use private key derivation in that case?